src/Compilers/Test/Core/TestableCompiler.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.VisualBasic;
+
+namespace Roslyn.Test.Utilities
+{
+    internal record struct TestableCompilerFile(string FilePath, TestableFile TestableFile)
+    {
+        internal List<byte> Contents => TestableFile.Contents;
+    }
+
+    internal enum BasicRuntimeOption
+    {
+        Include,
+        Exclude,
+        Embed,
+
+        // Consumer manually controls the vb runtime via the provided arguments
+        Manual,
+    }
+
+    internal sealed class TestableCompiler
+    {
+        internal static string RootDirectory => RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"q:\" : "/";
+        internal static BuildPaths StandardBuildPaths => new BuildPaths(
+            clientDir: Path.Combine(RootDirectory, "compiler"),
+            workingDir: Path.Combine(RootDirectory, "source"),
+            sdkDir: Path.Combine(RootDirectory, "sdk"),
+            tempDir: null);
+
+        internal CommonCompiler Compiler { get; }
+        internal TestableFileSystem FileSystem { get; }
+        internal BuildPaths BuildPaths { get; }
+
+        internal TestableCompiler(CommonCompiler compiler, TestableFileSystem fileSystem, BuildPaths buildPaths)
```

Comment:
>fileSystem [](http://example.com/codeflow?start=78&length=10)

Why are we passing in a `fileSystem` if it has to be the same as `compiler.FileSystem`? Can't we just use `compiler.FileSystem` then, and remove the extra parameter?

Reply:
The problem with the property is that it attaches *after* command line arguments are parsed where the ctor parameter has a chance to thread it through. It's long been a problem with this setup and one that I'm slowly working my waay trhough.